### PR TITLE
Create Heuristic Class and Particle Class

### PR DESCRIPTION
**Pull request by xjh666**

### CODE REVIEW

File: src/PlayerSkeleton.java [hunk 1]
```diff
@@ -25,3 +26,204 @@ public static void main(String[] args) {
 	}
 	
 }
+
+class Heuristics{
```
Comment (billstark) [inline src/PlayerSkeleton.java]: maybe can add a space before the bracket?

Comment (xjh666) [reply src/PlayerSkeleton.java]: done

File: src/PlayerSkeleton.java [hunk 2]
```diff
@@ -25,3 +26,204 @@ public static void main(String[] args) {
...
+	private double weightOfTotalHeight;
+	
+	private int rangeOfHeight;
+	
+	private double weightOfRangeOfHeight;
+	
+	
+	public Heuristics(State thisState, State nextState) {
+		this.lineCleared = nextState.getRowsCleared();
+
```
Comment (billstark) [inline src/PlayerSkeleton.java:104]: perhaps you could abstract out these operations into individual functions like `getNumberOfHoles(State state)`

File: src/PlayerSkeleton.java [hunk 3]
```diff
@@ -25,3 +26,204 @@ public static void main(String[] args) {
...
+			}
+		}
+		
+		this.totalHeight = 0;
+		for(int column = 0; column < State.COLS; column++) {
+			this.totalHeight += nextState.getTop()[column];
+		}
+		
+		this.rangeOfHeight = Arrays.stream(nextState.getTop()).max().getAsInt() -  Arrays.stream(nextState.getTop()).min().getAsInt();
+}
```
Comment (billstark) [inline src/PlayerSkeleton.java]: indentation here lol

File: src/PlayerSkeleton.java [hunk 4]
```diff
@@ -25,3 +26,204 @@ public static void main(String[] args) {
...
+	public Heuristics(State thisState, State nextState) {
+		this.lineCleared = nextState.getRowsCleared();
+		
+		this.numberOfHoles = 0;
+		this.totalWeightOfHoles = 0;
+		for(int column = 0; column < State.COLS; column++) {
+			for(int row = nextState.getTop()[column]-1; row >= 0; row--) {
+				if(nextState.getField()[row][column] == 0){
+					this.numberOfHoles++;
+					this.totalWeightOfHoles += (20 - row);
```
Comment (billstark) [inline src/PlayerSkeleton.java]: 20 is a magic number, maybe we could use State.ROWS - 1?

File: src/PlayerSkeleton.java [hunk 5]
```diff
@@ -25,3 +26,204 @@ public static void main(String[] args) {
 	}
 	
 }
+
```
Comment (billstark) [inline src/PlayerSkeleton.java]: One of the most important thing is that we might need to extract weights out. The Heuristic class just calculates the score for a specific state based on given weights. So as far as I am concern, the construction of Heuristic could be `Heuristics(State state, Double[] weights)` somethings like this

Comment (billstark) [reply src/PlayerSkeleton.java]: actually the `Double[] weights` would come from a class called `Particles` similar to the code that I shared with you yesterday. The `Particles` class will store the weight vector and use PSO algorithm to update the weights by iterations.

File: src/PlayerSkeleton.java [hunk 6]
```diff
@@ -25,3 +26,204 @@ public static void main(String[] args) {
...
+		}	
+		
+		this.sumOfAdjacentColumnHeightDifference = 0;
+		for(int column = 1; column < State.COLS; column++) {
+			this.sumOfAdjacentColumnHeightDifference += Math.abs(nextState.getTop()[column]-nextState.getTop()[column-1]);
+		}
+		
+		//the height of bottom gird 
+		this.landingHeight = State.ROWS;
+		for(int column = 0; column < State.COLS; column++) {
```
Comment (billstark) [inline src/PlayerSkeleton.java]: I am not really sure about this actually. how should landing height be calculated?

File: src/PlayerSkeleton.java (hunk 5 above)
Comment (billstark) [reply src/PlayerSkeleton.java]: For now my idea is that we can keep this Heuristics function, but just use it as a "calculator". We can make it a static class or singleton, then with weights or, `Particle` and `State` provide we could return the corresponding score.

File: src/PlayerSkeleton.java [hunk 7]
```diff
@@ -1,6 +1,13 @@
+import java.util.Arrays;
+import java.util.Random;
 
 public class PlayerSkeleton {
-
+	
+	public final static double INERTIA_WEIGHT = 0.72;
```
Comment (billstark) [inline src/PlayerSkeleton.java]: I think those should should be defined in `Particles`

**Pull request merged**